### PR TITLE
feat: integration test for incremental sync

### PR DESCRIPTION
**Pull request by vikash390**
# Description

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [ ] Scenario A
- [ ] Scenario B

# Screenshots or Recordings

### CODE REVIEW

File: utils/testutils/test_utils.go [hunk 1]
```diff
@@ -261,15 +282,30 @@ func (cfg *IntegrationTest) TestIntegration(t *testing.T) {
 							// 	`jq '(.selected_streams[][] | .normalization) = true' %s > /tmp/streams.json && mv /tmp/streams.json %s`,
 							// 	cfg.TestConfig.CatalogPath, cfg.TestConfig.CatalogPath,
 							// )
-							streamUpdateCmd := updateStreamsCommand(*cfg.TestConfig, cfg.Namespace, []string{currentTestTable}, true)
+							streamUpdateCmd := updateSelectedStreamsCommand(*cfg.TestConfig, cfg.Namespace, cfg.PartitionRegex, []string{currentTestTable}, true)
 							if code, out, err := utils.ExecCommand(ctx, c, streamUpdateCmd); err != nil || code != 0 {
 								return fmt.Errorf("failed to enable normalization in streams.json (%d): %s\n%s",
```
Comment (vaibhav-datazip) [inline utils/testutils/test_utils.go]: Update this error message as well

File: utils/testutils/test_utils.go [hunk 2]
```diff
@@ -306,30 +342,54 @@ func (cfg *IntegrationTest) TestIntegration(t *testing.T) {
...
+							cfg.ExecuteQuery(ctx, t, []string{currentTestTable}, "clean", false)
+							cfg.ExecuteQuery(ctx, t, []string{currentTestTable}, "add", false)
+
+							// Ensure normalization remains on for selected stream
+							streamUpdateCmd = updateSelectedStreamsCommand(*cfg.TestConfig, cfg.Namespace, cfg.PartitionRegex, []string{currentTestTable}, true)
+							if code, out, err := utils.ExecCommand(ctx, c, streamUpdateCmd); err != nil || code != 0 {
+								return fmt.Errorf("failed to enable normalization in streams.json for incremental (%d): %s\n%s",
+									code, err, out,
+								)
+							}
```
Comment (vaibhav-datazip) [inline utils/testutils/test_utils.go]: This is getting repeated here.

File: utils/testutils/test_utils.go [hunk 3]
```diff
@@ -306,30 +342,54 @@ func (cfg *IntegrationTest) TestIntegration(t *testing.T) {
...
+							if err := runSync(c, true, "", "r", cfg.ExpectedData); err != nil {
+								return err
+							}
+
+							// Delta incremental: add new rows and sync again
+							t.Log("Running Incremental - inserts")
+							cfg.ExecuteQuery(ctx, t, []string{currentTestTable}, "insert", false)
+							if err := runSync(c, true, "", "u", cfg.ExpectedData); err != nil {
+								return err
+							}
```
Comment (vaibhav-datazip) [inline utils/testutils/test_utils.go]: here as well a test cases array can be created and call that same for loop

File: utils/testutils/test_utils.go (hunk 1 above)
Comment (vikaxsh) [reply utils/testutils/test_utils.go]: what needs to update here?

File: utils/testutils/test_utils.go (hunk 2 above)
Comment (vikaxsh) [reply utils/testutils/test_utils.go]: removed

File: utils/testutils/test_utils.go (hunk 3 above)
Comment (vikaxsh) [reply utils/testutils/test_utils.go]: done

File: utils/testutils/test_utils.go [hunk 4]
```diff
@@ -340,7 +360,58 @@ func (cfg *IntegrationTest) TestIntegration(t *testing.T) {
...
+								},
+								{
+									name:       "insert",
+									setupQuery: "insert",
+									useState:   true,
+									opSymbol:   "u",
+									expected:   cfg.ExpectedData,
+								},
+							}
+
```
Comment (vaibhav-datazip) [inline utils/testutils/test_utils.go]: Can we not define these test cases along with other once and call it from that single for loop . instead of creating another array for incremental test cases.

File: drivers/postgres/internal/postgres_test.go [hunk 5]
```diff
@@ -18,6 +18,8 @@ func TestPostgresIntegration(t *testing.T) {
 		DataTypeSchema:     PostgresToIcebergSchema,
 		ExecuteQuery:       ExecuteQuery,
 		IcebergDB:          "postgres_postgres_public",
+		CursorField:        "col_bigserial",
+		PartitionRegex:     "/{col_bigserial,identity}",
 	}
```
Comment (vaibhav-datazip) [inline drivers/postgres/internal/postgres_test.go:23]: for postgres test, geometrical and array based datatypes should also be added .

Comment (vikaxsh) [reply drivers/postgres/internal/postgres_test.go:23]: we can create an issue for that , that is not part of this pr

File: utils/testutils/test_utils.go (hunk 4 above)
Comment (vikaxsh) [reply utils/testutils/test_utils.go]: since i have divided test cases into phase, before each phase we need to perform certain ops, for that keeping like that makes it simple, can add parquet test cases similarly for both cdc and incremental

File: drivers/postgres/internal/postgres_test.go (hunk 5 above)
Comment (vaibhav-datazip) [reply drivers/postgres/internal/postgres_test.go:23]: adding a few line change for incremental wont need a new PR , it can be done here.

The scope of this PR is to add incremental integration tests, you should add the test covering all the edge cases.

File: utils/testutils/test_utils.go (hunk 4 above)
Comment (vaibhav-datazip) [reply utils/testutils/test_utils.go]: You can check the PR #548 those extra steps are cleanly handled there, you can implement a similar thing in your PR

File: drivers/postgres/internal/postgres_test.go (hunk 5 above)
Comment (vikaxsh) [reply drivers/postgres/internal/postgres_test.go:23]: geometrical and array based datatypes are not part of incremental integration test

Comment (vaibhav-datazip) [reply drivers/postgres/internal/postgres_test.go:23]: don't add them as cursor field just add the datatypes in table which is getting synced.

for now array type will give error, so you may skip that as that is being fixed in PR #646

File: utils/testutils/test_utils.go (hunk 4 above)
Comment (vikaxsh) [reply utils/testutils/test_utils.go]: Adding a reset check inside the test array is incorrect it is out of scope of test case array should only contains test cases info,  that logic needs to be handled separately.

File: utils/testutils/test_utils.go [hunk 6]
```diff
@@ -177,6 +177,187 @@ func GetBackfillStreamsFromCDC(cdcStreams []string) []string {
...
+	testTable string,
+) error {
+	t.Log("Starting Phase B: Full load + Incremental tests")
+
+	// Reset table to isolate incremental scenario
+	cfg.ExecuteQuery(ctx, t, []string{testTable}, "drop", false)
+	cfg.ExecuteQuery(ctx, t, []string{testTable}, "create", false)
+	cfg.ExecuteQuery(ctx, t, []string{testTable}, "clean", false)
+	cfg.ExecuteQuery(ctx, t, []string{testTable}, "add", false)
+
```
Comment (vaibhav-datazip) [inline utils/testutils/test_utils.go:307]: abstract this out in a function, which will be called from anywhere and will clean the db and re add the data

File: utils/testutils/test_utils.go [hunk 7]
```diff
@@ -177,6 +177,187 @@ func GetBackfillStreamsFromCDC(cdcStreams []string) []string {
...
+			opSymbol:  "r",
+			expected:  cfg.ExpectedData,
+		},
+		{
+			name:      "insert",
+			operation: "insert",
+			useState:  true,
+			opSymbol:  "u",
+			expected:  cfg.ExpectedData,
+		},
```
Comment (vaibhav-datazip) [inline utils/testutils/test_utils.go:337]: we can also add an update case as well, also incremental should be tested for various types of cursor values

like 
- string 
- time 
- timestamp 
- timestamp with timezone
- datetime 
- float 
- int etc

File: utils/testutils/test_utils.go (hunk 6 above)
Comment (vikaxsh) [reply utils/testutils/test_utils.go:307]: this is already abstracted, calling functions only

File: utils/testutils/test_utils.go (hunk 7 above)
Comment (vikaxsh) [reply utils/testutils/test_utils.go:337]: we can create issues for adding multiple cursor fields and test them

Comment (vaibhav-datazip) [reply utils/testutils/test_utils.go:337]: please add a todo for this then

File: utils/testutils/test_utils.go (hunk 6 above)
Comment (vaibhav-datazip) [reply utils/testutils/test_utils.go:307]: currently its getting called from just this function , will abstract it out later if its called from other functions as well
